add RDMFT(Reduced Density Matrix Functional Theory) code （Useful Information: for those who what to contribute codes to ABACUS, this is a good example)

JGHan7 opened the pull request:
RDMFT is an ab-initio method.

I have implemented some code for rdmft in ABACUS. As long as the default value "ks" (instead, "rdmft") of the parameter ab_initio_type is used in INPUT, it will not affect the normal function of ABACUS.

![image](https://github.com/user-attachments/assets/17546a90-277f-48a7-b4b9-30ebe3dd4e27)

JGHan7 committed:
in last commit, 2 processor has 'malloc(): memory corruption:' because I do three for() to realize gemm and it doesn't support many processors...

JGHan7 committed:
first push with rdmft writed in abacus-v3.4.4

JGHan7 committed:
update README.md

JGHan7 committed:
update README.md again

JGHan7 committed:
find all parameters used in Veff_rdmft (for hartree and local, XC still has two parameters don't pass), and pass in rdmft_cal() successfully

JGHan7 committed:
use class Veff_rdmft get correct Ehartree and Eexx, but Etotal is wrong

JGHan7 committed:
all energy terms correct, ETV_RDMFT = E_one_elec, Ehartree_RDMFT = E_Hartree, EXC_RDMFT = E_exx, but my Etotal doesn't include ion energy terms, etc.

JGHan7 committed:
z/dgemm_() is used correctly, next step to use pz/dgemm_()

JGHan7 committed:
just one processor is correct, 2 processors have some problem and maybe the error is occurred in processor 2

JGHan7 committed:
2 processors is wrong but 6 processors is correct...

JGHan7 committed:
check 2 processors error

JGHan7 committed:
update, realize print Dmn in different processors

JGHan7 committed:
get successful result in different numbers of processors, and the printout has not been deleted yet

JGHan7 committed:
get the sum of different processors, that is, get the true total energy of each term

JGHan7 committed:
delete some printout and next step is replace the pz/dgemm_() used in psiDotPsi()

JGHan7 committed:
add some comments

JGHan7 committed:
do some little modifications to make the code more concise

JGHan7 committed:
just save

JGHan7 committed:
just save

JGHan7 committed:
last save before realizing gamma_only calculation

JGHan7 committed:
gamma_only calculation with errors

JGHan7 committed:
save with errors in gamma_only

JGHan7 committed:
gamma_only realization with errors in E_T+nonlocal

JGHan7 committed:
gamma_only successful with one processors, but multi-processor errors

JGHan7 committed:
support g(wg)*H_wfc and g(wg)*wfcHwfc with different g(wg) which is provided by HF_XCfunc or power_XCfunc

JGHan7 committed:
get the special DM used in constructing V_XC, and can be used for split_m2D_ktoR()

JGHan7 committed:
get the special DM used in constructing V_XC and run successfully (not in numerica meaning)

JGHan7 committed:
modify the namespace of rdmft_test files from hamilt to rdmft

JGHan7 committed:
error with using spilt_m2D_ktoR()

JGHan7 committed:
successfully get Ds_XC, next step is update abacus to have the new interface

JGHan7 committed:
merge with newest abacus-v3.5.0

JGHan7 committed:
successfully run rdmft with power functional in multi-k calculation, have tested alpha=1.0, 0.95, 0.5

JGHan7 committed:
do some little modifications to make the code more concise

JGHan7 committed:
fix interface provided by class OperatorEXX but having errors when do V_XC.contributeHk(ik)

JGHan7 committed:
print data of tensor from Ds_XC

JGHan7 committed:
just save

JGHan7 committed:
check the length and data of LM->Hexxc and Vxc_fromRI_c.Hexxs

JGHan7 committed:
have modified op_exx_lcao.h/.cpp, RI_2D_Comm.h/.cpp files, delete the modifying in the future

JGHan7 committed:
save

JGHan7 committed:
successfullu add power XC_functional and run correctly with Si soilds in 6 nodes and 6 processors

JGHan7 committed:
Merge branch 'develop' of https://gitee.com/deepmodeling/abacus-develop into rdmft

JGHan7 committed:
successfully realize power functional with testing Si solids, and some printout need to be deleted

JGHan7 committed:
Print: last push before delete many printout

JGHan7 committed:
delete modifying in source code

JGHan7 committed:
Refactor: first add class RDMFT

JGHan7 committed:
save

JGHan7 committed:
get DM used for updating charge/rho

JGHan7 committed:
try to initial DMR of class DensityMatrix

JGHan7 committed:
save before modifying gint.cpp

JGHan7 committed:
update charge using wg&wfc which provided by me successfully

JGHan7 committed:
delete some comment

JGHan7 committed:
last save before transfer wg -> natural occupation numbers (occupation_num)

JGHan7 committed:
successfully transfer wg -> occ_number and get correct result in Muller_XC

JGHan7 committed:
modifying some variable names, function names and comments (for wg -> occNum)

JGHan7 committed:
add module_rdmft and copy code here working successfully

JGHan7 committed:
delete rdmft_test.h/.cpp files in module_hamilt_lcao/... and modify corresponding CMakeLists.txt

JGHan7 committed:
add rdmft.h, rdmft.cpp, rdmft_tools.cpp files to create class RDMFT

JGHan7 committed:
just save. In the process of factor function rdmft_cal() -> class RDMFT

JGHan7 committed:
modify rdmft_solver in esolver_ks_lcao.h/.cpp

JGHan7 committed:
push with compile error in cal_dm.h( psiMulPsiMpi() function with the same name is wrong )

JGHan7 committed:
add class rdmft and do rdmft_solver.init() successfully, next step is init V_TV, V_XC...

JGHan7 committed:
successfully move functions of class RDMFT to rdmft.cpp

JGHan7 committed:
just save

JGHan7 committed:
just save

JGHan7 committed:
factor function rdmft_cal() -> class RDMFT and with running error

JGHan7 committed:
just save

JGHan7 committed:
Factor: transfer function rdmft_cal() -> class RDMFT and run in afterscf() successfully

JGHan7 committed:
just save

JGHan7 committed:
now get V_local in eiter, next step move it to istep, i.e. together with T_nonlocal

JGHan7 committed:
move the obtaining of V_local from elec-step to ion-step

JGHan7 committed:
successfully find where to do update_ion()

JGHan7 committed:
class RDMFT get correct E_TV and E_hartree, but E_XC=0.0

JGHan7 committed:
Factor: complete func rdmft_cal() -> class RDMFT and find where to use the RDMFT class object rdmft_solver appropriately

JGHan7 committed:
save

JGHan7 committed:
save for debug

JGHan7 committed:
Fix: fix bug in HkPsi(), transfer 'T'->'C'

JGHan7 committed:
save

JGHan7 committed:
Fix: fix nk_total in class RDMFT depend on the Symmetry::symm_flag

JGHan7 committed:
save

JGHan7 committed:
test different branch

JGHan7 committed:
add Dell_abaInstall.sh for Dell_server

JGHan7 committed:
modify module_rdmft to latest version

JGHan7 committed:
just save for updating branch rdmft

JGHan7 committed:
update module_xc for rdmft(WP22,CWP22 and so on)

PeizeLin committed:
Fix memory bug in XC_Functional_Libxc::cal_gdr()

JGHan7 committed:
merge to abacus-v3.7.5, but don't debug(many, because many functions and interfaces have changed)

PeizeLin committed:
Refactor XC_Functional_Libxc::convert_vtxc_v()

JGHan7 committed:
merge to abacus-v3.7850 but don't debug(many, because many functions and interfaces have changed). The scaling_factor_xc issue is not handled

JGHan7 committed:
merge to abacus-v3.8.0 but don't debug(many, because many functions and interfaces have changed). The scaling_factor_xc issue is not handled

JGHan7 committed:
Merge branch 'convert_vtxc_v' of https://gitee.com/linpeize/abacus-develop into rdmft

JGHan7 committed:
merge to abacus-v3.8.0 and add scaling_factor in libxc_cpp files

JGHan7 committed:
merge to abacus-v3.8.0 and debug is not finished

JGHan7 committed:
just save

JGHan7 committed:
just save

JGHan7 committed:
after merging into abacus-v3.8.0, the first compilation passed, but the program has not been run yet

JGHan7 committed:
after merging to abacus-v3.8.0, the program ran successfully for the first time, but the call of V_exx was not correct.

JGHan7 committed:
just save

JGHan7 committed:
just save

JGHan7 committed:
just save

JGHan7 committed:
the Vexx may have error when symmetry==1, skip it and continue do my own code

JGHan7 committed:
add parameters used in rdmft in input_parameters.h and so on

JGHan7 committed:
merge all rdmft-code to abacus-v3.8.0, next step is debug and test

JGHan7 committed:
debug now ,some energy is little different from abacus

JGHan7 committed:
Merge branch 'develop' of https://gitee.com/deepmodeling/abacus-develop into rdmft

JGHan7 committed:
merge to the latest aABACUS

JGHan7 committed:
Fix the bug caused by merging the latest ABAUCS, need to test other bugs such as wp22, cwp22 and the difference of energy

JGHan7 committed:
merge to ABACUS-v3.8.1, 2024-10-22

JGHan7 committed:
just save

JGHan7 committed:
add XC-functional BLYP_LIBXC and BLYP_LR to test

JGHan7 committed:
stop tracking abaInstall_HZWpara.sh

JGHan7 committed:
Merge remote-tracking branch 'upstream/develop' into rdmft

JGHan7 committed:
just save

JGHan7 committed:
just save

JGHan7 committed:
save

JGHan7 committed:
just save

JGHan7 committed:
modify cmake/FindELPA.cmake

JGHan7 committed:
save

JGHan7 committed:
add RDMFT(Reduced Density Matrix Functional Theory) code

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

maki49 commented:
Excellent work!
You can do the following steps at first. After that, I will give you some suggestion about how to refactor your code.
- Remove the useless files (like "backward.hpp")
- Compile the code successfully on your local machine based on the latest develop branch
- Add some integrate test cases ( see [how-to-add-integrate-tests-(by-dyzheng).pdf](https://github.com/user-attachments/files/17478817/how-to-add-integrate-tests-.by-dyzheng.pdf)) to prevent the code from being destroyed by the subsequent modification

kirk0830 commented on docs/advanced/input_files/input-main.md:
```diff
@@ -3992,4 +3995,20 @@ The output files are `OUT.${suffix}/Excitation_Energy.dat` and `OUT.${suffix}/Ex
+
+ab-initio methods and the xc-functional parameters used in RDMFT.
+
+### ab_initio_type
```
Hi, there is another keyword `esolver_type`, and `ksdft` is a possible value of it, could you check if `rdmft` can be a param of it?

JGHan7 replied:
theoretically, rdmft can be a value of esolver_type.

But there are too many places in the program that involve esolver_type, which makes me think that it is dangerous to use esolver_type to express rdmft   (when calculating rdmft, we usually need to run one or several steps of ksdft calculation to get the initial value, so when running rdmft, we must ensure that the ksdft program can run completely normally)

So I temporarily added the parameter ab_initio_type

JGHan7 replied:
Thank you for your opinion, senior sister.

In fact, the theoretical status of RDMFT and DFT is the same, but when we apply RDMFT in practice, we can use the results of one or several steps of KS as the initial value (of course, we can also use a random 1-RDM as the initial value), which requires KSDFT to run completely in the actual program.

However, in the current RDMFT implementation, I am still writing the program for variational optimization to the 1-RDM of the ground state (the optimization program of the cooperating teacher is not fully written yet), so it seems that the code I pushed up is doing some "repetitive" calculations without practical significance except for using a few new functionals.

Professor Ren meant that, because ABACUS is often refactored, I need to spend a long time to rewrite the correct rdmft code in the old version into the new version every once in a while, which is a waste of time and has no practical significance for the development of new methods. Therefore, it is recommended that I merge the rdmft code into the developer branch in time, but not release it in the user-oriented version for the time being (as long as my code itself does not affect the normal code of ABACUS, for example, when the ab_initio_type of INPUT is the default "ksdft", my code has no effect on other calculations of ABACUS), so the current rdmft module is still considered a production version.

The physical quantities that rdmft temporarily expects to output are the kinetic energy, total energy and 1-RDM of the system in the ground state (by the way, where is this document)

JGHan7 replied:
Thank you for your opinion, senior sister.

In fact, the theoretical status of RDMFT and DFT is the same, but when we apply RDMFT in practice, we can use the results of one or several steps of KS as the initial value (of course, we can also use a random 1-RDM as the initial value), which requires KSDFT to run completely in the actual program.

However, in the current RDMFT implementation, I am still writing the program for variational optimization to the 1-RDM of the ground state (the optimization program of the cooperating teacher is not fully written yet), so it seems that the code I pushed up is doing some "repetitive" calculations without practical significance except for using a few new functionals.

Professor Ren meant that, because ABACUS is often refactored, I need to spend a long time to rewrite the correct rdmft code in the old version into the new version every once in a while, which is a waste of time and has no practical significance for the development of new methods. Therefore, it is recommended that I merge the rdmft code into the developer branch in time, but not release it in the user-oriented version(as long as my code itself does not affect the normal code of ABACUS, for example, when the ab_initio_type of INPUT is the default "ksdft", my code has no effect on other calculations of ABACUS), so the current rdmft module is still considered a production version.

The physical quantities that rdmft temporarily expects to output are the kinetic energy, total energy and 1-RDM of the system in the ground state (by the way, where is this document)

maki49 replied:
I see. I mean such information can appear in the parameter description in this .md file.
> The physical quantities that rdmft temporarily expects to output are the kinetic energy, total energy and 1-RDM of the system in the ground state (by the way, where is this document)

JGHan7 replied:
OK，I'll add it, thank you

mohanchen replied:
I would suggest changing " ab_initio_type" to "rdmft" and set its type to bool, and make this variable controlled by yourself. Do not invent any ambiguous variable for other developers to think they can do something new with this variable.

kirk0830 commented on lib64/backward/BackwardConfig.cmake:
```diff
@@ -0,0 +1,277 @@
+#
+# BackwardMacros.cmake
+# Copyright 2013 Google Inc. All Rights Reserved.
```
do you have some documents illustrating your newly added dependencies?

JGHan7 replied:
In fact, I don't know what this file is, I will delete lib64/ and include/

kirk0830 commented on source/module_esolver/esolver_ks.cpp:
```diff
@@ -623,6 +630,31 @@ void ESolver_KS<T, Device>::runner(const int istep, UnitCell& ucell)
+
+                //initialize the gradients of Etotal on wg and wfc, and set all elements to 0. 
+                ModuleBase::matrix E_gradient_occNum(this->pelec->wg.nr, this->pelec->wg.nc, true);
+                psi::Psi<T> E_gradient_wfc(this->psi->get_nk(), this->psi->get_nbands(), this->psi->get_nbasis()); 
```
would the name of variable be more precise if it were `dedpsi` instead? Plus, although might not be critical, but is it necessary to allocate the memory for all kpoints in-one-shot?

JGHan7 replied:
ok，I will use dE_dwfc to replace it.

Because my program obtains all k points at once when calculating the gradient, I also do this here.

In addition, E_gradient_occNum and E_gradient_wfc appear here only to verify the correctness of the program. After I write the optimized program, I will delete these two variables from here (the code in esolver_ks.cpp is only for getting the initial value of rdmft calculation after one or several steps of ksdft)

kirk0830 commented on source/module_esolver/esolver_ks.cpp:
```diff
@@ -623,6 +630,31 @@ void ESolver_KS<T, Device>::runner(const int istep, UnitCell& ucell)
+                this->update_elec_rdmft(occ_number_ks, *(this->psi));
+
+                //initialize the gradients of Etotal on wg and wfc, and set all elements to 0. 
+                ModuleBase::matrix E_gradient_occNum(this->pelec->wg.nr, this->pelec->wg.nc, true);
```
would the name of variable be more precise if it were `dedfnk`, along with an annotation `// fnk: occupation of band n, at k` instead?

JGHan7 replied:
dEdOccNum may be acceptable,

but occNum (natural occupation number) is an important physical quantity in rdmft, namely the eigenvalue of the single-particle reduced density matrix, which should be distinguished from fnk in dft (although I use fnk of ksdft as the initial value of the occupation number)

kirk0830 commented on source/module_esolver/esolver_ks.h:
```diff
@@ -10,6 +10,9 @@
 #include "module_io/cal_test.h"
 #include "module_psi/psi.h"
 
+// // add by jghan for rdmft calculation
+// #include "module_rdmft/rdmft.h"
+
```
seems it is not used, so I remind you remove it when your PR is ready for merging

JGHan7 replied:
I will delete it

kirk0830 commented on source/module_esolver/esolver_ks.h:
```diff
@@ -72,6 +75,10 @@ class ESolver_KS : public ESolver_FP
         //! <Temporary> It should be replaced by a function in Hamilt Class
 		virtual void update_pot(const int istep, const int iter) {};
 
+        virtual double Run_rdmft(ModuleBase::matrix& E_gradient_occNum, psi::Psi<T>& E_gradient_wfc) { return 0.0; };   // add by jghan, 2024-03-16
```
We have a convention of naming classes, functions and variables. For classes, it is recommended to use something like `ReducedDensityMatrixFunctionalTheory`, for functions and variables, use `reduced_density_matrix_functional_theory` or abbr of it like `rdmft_*`. The function name `update_elec_rdmft` is a good practice. :)

JGHan7 replied:
ok, I  will use run_rdmft() to replace

kirk0830 commented on source/module_esolver/esolver_ks_lcao.cpp:
```diff
@@ -55,6 +55,12 @@
 #include "module_io/write_wfc_nao.h"
 
+// test RDMFT
+#include "module_rdmft/rdmft.h"
+#include "module_rdmft/rdmft_tools.h"
+#include "module_rdmft/rdmft_test.h"
```
there might be a not nice dependency between functionas declared in rdmft, _tools and _test. The concept is, you only need to expose an interface to rdmft in esolver_ks_lcao, which is a derived driver. @maki49 Could you help him design a littble bit?

JGHan7 replied:
Yes, that's exactly what you understood, and may need your help to complete it better

kirk0830 commented on source/module_esolver/esolver_ks_lcao.cpp:
```diff
@@ -1178,6 +1196,171 @@ void ESolver_KS_LCAO<TK, TR>::after_scf(const int istep)
+    // double E_TV_Hartree_XC = this->pelec->f_en.deband + this->pelec->f_en.eband 
+    //                             + this->pelec->f_en.hartree_energy + this->pelec->f_en.etxc - this->pelec->f_en.etxcc;
+    // std::cout << std::fixed << std::setprecision(10) << "\n\n******\nE(one_elec + Hartree + EXX) by KS-DFT:  " << E_TV_Hartree_EXX
+    //             << "\n\nE(one_elec + Hartree + XC) by KS-DFT:  " << E_TV_Hartree_XC << "\nExc_KS:         " << E_xc_KS << "\n******\n\n" << std::endl;
+
+    // ModuleBase::TITLE("RDMFT", "E & Egradient");
+    // ModuleBase::timer::tick("RDMFT", "E & Egradient");
+
+    // // // test class rdmft
+    // // rdmft_solver.update_elec(occ_number, *(this->psi));
+    // // this->Run_rdmft(E_gradient_occNum, E_gradient_wfc);
+    // // std::cout << "\nrdmft_solver: " << "0.0000" << std::endl;
+
+    // // double minus = 0.0;
+
+    // // // delete in the future
+    // // std::cout << "\n\n occ_number(ik, inband): " << std::endl;
+    // // for(int ik=0; ik < occ_number.nr; ++ik)
+    // // {
+    // //     for(int inb=0; inb < occ_number.nc; ++inb) std::cout << occ_number(ik, inb) << " ";
+    // //     std::cout << "\n" << std::endl;
+    // // }
+
+    // // std::cout << "\n\n wg(ik, inband): " << std::endl;
+    // // for(int ik=0; ik < occ_number.nr; ++ik)
+    // // {
+    // //     for(int inb=0; inb < occ_number.nc; ++inb)
+    // //     {
+    // //         std::cout << this->pelec->wg(ik, inb) << " ";
+    // //         minus += this->pelec->wg(ik, inb) - occ_number(ik, inb)*this->kv.wk[ik];
+    // //     }
+    // //     std::cout << "\n" << std::endl;
+    // // }
+
+    // // if( std::abs(minus) < 1e-12 ) std::cout << "\n\nminus<1e-12, minus: " << minus << std::endl;
+    // // else if( std::abs(minus) < 1e-16 ) std::cout << "\n\nminus<1e-16, minus: " << minus << std::endl;
+    // // else {std::cout << "\n\nminus is big, minus: " << minus << std::endl;}
+
+    // ModuleBase::timer::tick("RDMFT", "E & Egradient");
+
```
Remember to remove them in your production code, thanks

JGHan7 replied:
I'm very sorry, I have forgotten this code...
I'll delete it now

kirk0830 reviewed (commented):
I have reviewed part of this PR, will complete my review later, thanks for your contribution. 
:)

mohanchen committed:
Merge branch 'develop' into rdmft_PR

JGHan7 committed:
delete lib64/ and include/

JGHan7 committed:
delete some production code and modify some variable/function names

mohanchen committed:
Merge branch 'develop' into rdmft_PR

JGHan7 committed:
save

JGHan7 committed:
Merge branch 'develop' of https://gitee.com/deepmodeling/abacus-develop into rdmft

JGHan7 committed:
merge to the latest abacus

JGHan7 committed:
merge and modify rdmft_PR branch

JGHan7 committed:
Merge branch 'rdmft_PR' of github.com:JGHan7/abacus-develop into rdmft_PR

JGHan7 committed:
just save

JGHan7 committed:
add tests for rdmft

JGHan7 committed:
Merge branch 'develop' of https://gitee.com/deepmodeling/abacus-develop into rdmft

JGHan7 committed:
change branch from rdmft to rdmft_PR, prepare for PR

JGHan7 committed:
recover accidentally deleted files

maki49 commented on source/module_esolver/esolver_ks.cpp:
```diff
@@ -623,6 +630,36 @@ void ESolver_KS<T, Device>::runner(const int istep, UnitCell& ucell)
 #endif //__RAPIDJSON
 
+        // 12.5) rdmft, add by jghan 2024-04-08/2024-10-09
+        if ( PARAM.inp.ab_initio_type == "rdmft" )
```
I guess this part can be put into `ESolver_KS_LCAO::iter_finish`, where you can directly call `rdmft_solver.run` and `rdmft_solver.update_elec`. Then the wrapper functions in esolver_ks.h can be removed.

JGHan7 replied:
OK，I'll do it

maki49 commented on source/module_rdmft/rdmft.h:
```diff
@@ -0,0 +1,254 @@
+    // Or we can use rdmft_solver.wfc/occ_number directly when optimizing, so that the update_elec() function does not require parameters.
+    void update_elec(const ModuleBase::matrix& occ_number_in, const psi::Psi<TK>& wfc_in, const Charge* charge_in = nullptr);
+
+    // update occ_number for optimization algorithms that depend on Hamilton
+    void update_occNumber(const ModuleBase::matrix& occ_number_in);
+
+    // // update occ_number for optimization algorithms that depend on Hamilton
+    // void update_wg(const ModuleBase::matrix& wg_in);
+
+    // get the total Hamilton in k-space
+    void cal_Hk_Hpsi();
```
Seems these two can be private.

JGHan7 replied:
OK, I will mark cal_Hk_Hpsi() as private. 

update_occNumber() is an interface reserved for the optimizer and is not used yet, public should be more appropriate.

maki49 commented on source/module_esolver/esolver_ks_lcao.cpp:
```diff
@@ -1178,6 +1196,30 @@ void ESolver_KS_LCAO<TK, TR>::after_scf(const int istep)
     ModuleBase::timer::tick("ESolver_KS_LCAO", "out_deepks_labels");
 #endif
 
+    /******** test RDMFT *********/
+
+    // rdmft, added by jghan, 2024-10-17
+    if ( PARAM.inp.ab_initio_type == "rdmft" )
+    {
+        ModuleBase::matrix occ_number_ks(this->pelec->wg);
+        for(int ik=0; ik < occ_number_ks.nr; ++ik)
+        {
+            for(int inb=0; inb < occ_number_ks.nc; ++inb) { occ_number_ks(ik, inb) /= this->kv.wk[ik];
+}
+        } 
+        this->update_elec_rdmft(occ_number_ks, *(this->psi));
+
+        //initialize the gradients of Etotal on occupation numbers and wfc, and set all elements to 0. 
+        ModuleBase::matrix dE_dOccNum(this->pelec->wg.nr, this->pelec->wg.nc, true);
+        psi::Psi<TK> dE_dWfc(this->psi->get_nk(), this->psi->get_nbands(), this->psi->get_nbasis()); 
+        dE_dWfc.zero_out();
+
+        double Etotal_RDMFT = this->run_rdmft(dE_dOccNum, dE_dWfc);
+    }
+
+    /******** test RDMFT *********/
```
The call here seems no different from the call in each iter (and gives the same result as the last iter in your test case), is it really needed?

JGHan7 replied:
Lines 1211 to 1217 are not needed (as mentioned before, they are production code and will actually be used elsewhere). 

They are just to ensure that there are more results in the test and to ensure that my code is not broken during refactoring.

I will delete those meaningless codes when releasing

maki49 commented on source/module_rdmft/rdmft.cpp:
```diff
@@ -0,0 +1,979 @@
+    std::cout << "\netxc:  " << etxc << "\nvtxc:  " << vtxc << "\n";
+    std::cout << "\nE_deband_KS:  " << E_deband_KS << "\nE_deband_harris_KS:  " << E_deband_harris_KS << "\n\n" << std::endl;
+
+    if( 1 )
+    {
+        // GlobalV::ofs_running << std::setprecision(12);
+        // GlobalV::ofs_running << std::setiosflags(std::ios::right);
+        GlobalV::ofs_running << std::fixed << std::setprecision(10)
+                << "\n******\nE(TV + Hartree + XC) by RDMFT:   " << E_RDMFT[3] 
+                << "\n\nE_TV_RDMFT:      " << E_RDMFT[0] 
+                << "\nE_hartree_RDMFT: " << E_RDMFT[1] 
+                << "\nExc_" << XC_func_rdmft << "_RDMFT:    " << E_RDMFT[2] 
+                << "\nE_Ewald:         " << E_Ewald
+                << "\nE_entropy(-TS):  " << E_entropy 
+                << "\nE_descf:         " << E_descf 
+                << "\n\nEtotal_RDMFT:    " << Etotal 
+                << "\n\nExc_ksdft:       " << E_xc_KS 
+                << "\nE_exx_ksdft:     " << E_exx_KS 
+                <<"\n******\n" << std::endl;
+    }
```
Seems duplicated?

JGHan7 replied:
Yes, it's a repeat... I can comment out the output on the terminal screen (or "job.log") first.

maki49 commented:
Please add your new directories and "*.o" of new cpp files to [source/Makefile.Objects](https://github.com/deepmodeling/abacus-develop/blob/develop/source/Makefile.Objects) to make your code pass the Makefile build test.

maki49 commented on source/module_rdmft/rdmft.h:
```diff
@@ -0,0 +1,254 @@
+    hamilt::OperatorLCAO<TK, TR>* V_hartree_XC = nullptr;
+    // bool get_V_local_temp = true;
+
+    Exx_LRI<double>* Vxc_fromRI_d = nullptr;
+    Exx_LRI<std::complex<double>>* Vxc_fromRI_c = nullptr;
```
Oops, you need to add new `Exx_LRI` objects here...

Seems they share the same `ccp_type` and `ccp_rmesh_times` as those in esolver_ks_lcao.h, namely `exx_lri_double` and `exx_lri_complex`. The only difference is its `Hexxs` member calculted with different density matrices by `cal_exx_elec`, right? But the new objects here double the memory cost of ABFs, two-center tables and RI- C, V tensors...

I think class `Exx_LRI` needs some refactor.

JGHan7 replied:
yes,  the only difference is its Hexxs member calculted with different density matrices by cal_exx_elec.

and I give ccp_rmesh_times as shown below (ccp_type is similar)
![ccp_rmesh_times_forRDMFT](https://github.com/user-attachments/assets/cefa9a6a-c913-42e9-ab55-772762347eff)

JGHan7 committed:
move rdmft related functions and their calls from esolver_ks to esolver_ks_lcao, and so on

JGHan7 committed:
Merge branch 'develop' of https://gitee.com/deepmodeling/abacus-develop into rdmft

JGHan7 committed:
save

JGHan7 committed:
save

JGHan7 committed:
save

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

JGHan7 committed:
add '#ifdef __MPI', delete run_rdmft() and so on

JGHan7 committed:
just save

JGHan7 committed:
prepare for PR

JGHan7 committed:
prepare for PR

JGHan7 committed:
Merge branch 'rdmft' into rdmft_PR

JGHan7 committed:
save

JGHan7 committed:
debug

JGHan7 committed:
add if(ENABLE_LCAO) in rdmft/CMakeLists.txt

JGHan7 committed:
Merge branch 'develop' of https://gitee.com/deepmodeling/abacus-develop into rdmft

JGHan7 committed:
save

JGHan7 committed:
Merge branch 'rdmft' of gitlab.com:JGHan7/rdmft-abacus into rdmft

JGHan7 committed:
save

JGHan7 committed:
prepare for PR

maki49 committed:
fix compile problem

JGHan7 committed:
Merge pull request #1 from maki49/rdmft_PR

fix compile problem

JGHan7 committed:
save

JGHan7 committed:
modify the error in module_io/test/read_input_ptest.cpp

JGHan7 committed:
prepare for PR

maki49 commented:
@mohanchen I think it's ready to merge.

JGHan7 committed:
save

JGHan7 committed:
add symmetry calculation of exx in rdmft

JGHan7 committed:
add test in esolver_ks_lcao.cpp, function after_scf(): update exx to test the exx_energy get by rdmft is exactly equal to exx_energy by abacus

JGHan7 committed:
delete rdmft_test.h/.cpp

JGHan7 committed:
delete the test for exx in esolver_ks_lcao.cpp, after_scf(

JGHan7 committed:
update test with symmetry, tests/integrate/1001_NO_Si2_dzp_rdmft

JGHan7 committed:
prepare for PR

JGHan7 committed:
save

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

mohanchen commented on source/module_esolver/esolver_ks_lcao.h:
```diff
@@ -78,6 +81,8 @@ class ESolver_KS_LCAO : public ESolver_KS<TK> {
 
     TwoCenterBundle two_center_bundle_;
 
+    rdmft::RDMFT<TK, TR> rdmft_solver;  // added by jghan for rdmft calculation, 2024-03-16
```
I would suggest not adding any data in esolver_ks_lcao.h

maki49 replied:
Sorry that I cannot find anywhere better to put this object, because its life span is a whole ion-step. Do you have any suggestions?

mohanchen replied:
I would not suggest writing code in this way, because: for example, if the initialization of rdmft causes a segmentation fault, it will affect the original LCAO developers. A tmp solution may be: inherit esolver_ks_lcao and generate a new esolver: esolver_ks_rdmft class, then you can add this variable in that esolver_ks_rdmft? In that case, the original LCAO will not depend on any details about rdmft.

mohanchen replied:
Therefore, the rdmft can be a esolver, it is fine rdmft needs esolver_ks_lcao because rdmft inherits esolver_ks_lcao

maki49 replied:
I agree, but not this PR (it is just a initial-guess generation).  
The true rdmft esolver will come in with the optimization algirithm in the future, which will inherit `ESolver_KS_LCAO`.

JGHan7 replied:
Thank you very much for Mr. Chen's suggestion.

Yes, inheriting the esolver_ks_lcao class is a better way to write it, but I need to explain that as long as the parameter ab_initio_type uses the default value, there is only one class object declaration for RDMFT. Its initialization, calling, etc. will not happen, that is, it has no effect on the ABACUS program.

Another reason is that Mr. Luo Kai, who cooperated with us to write the direct minimization module, has inherited the esolver_ks_lcao class in his code. If I inherit it here, our two programs will be repeated and conflicting, which will cause trouble for future docking (more importantly, it will cause duplication of our work and waste of time. In the future, someone will always delete the inheritance of the esolver_ks_lcao class and re-modify the code)

Back to the beginning, my original intention of merging such an incomplete rdmft code into the developer branch was to avoid rewriting the working code into the latest ABACUS again and again, which is a waste of time and wears out people's enthusiasm. (Mr. Ren means that after rdmft is merged into the developer branch, the person who does the refactoring needs to actively modify the rdmft code to make it work)

Thank you again for your advice. I can guarantee that when we actually release the rdmft function to users, we will use a method such as inheriting the esolver_ks_lcao class to merge ABACUS. Do you think this is okay?

kirk0830 replied:
@maki49 is the RDMFT still under framework of KS-DFT? I am a littble confused about the conceptual relationship between the KS-DFT and RDMFT

JGHan7 replied:
In my new development branch, this has been deleted. My new code inherits ESolver_KS_LCAO (regardless of how to generate the initial value, what I really want to inherit is ESolver_LCAO, but we don't have it).

In theory, you can understand that RDMFT and KS-DFT have nothing to do with each other, but in the ABACUS code framework, many of the quantities I need come from ESolver_KS_LCAO

mohanchen commented on source/module_io/input_conv.cpp:
```diff
@@ -319,7 +319,15 @@ void Input_Conv::Convert()
+            || dft_functional_lower == "wp22" 
+            || dft_functional_lower == "cwp22" ) // added by jghan, 2024-07-07
+        {
+            GlobalC::restart.info_save.save_charge = true;
```
Can we not using GlobalC::restart??

maki49 replied:
I will refactor this in another PR.

mohanchen commented on source/module_parameter/input_parameter.h:
```diff
@@ -586,5 +586,12 @@ struct Input_para
+
+    // ==============   #Parameters (21.RDMFT) =====================
+    // RDMFT    jghan added on 2024-07-06
+    std::string ab_initio_type = "ksdft";       // "rdmft", reduced density matrix funcional theory
```
not sure we should add this parameter

maki49 replied:
Personally speaking, it's not a good idea to add an option under `esolver_type`, because this PR does not add a new esolver type... In contrast, adding a new parameter helps to reduce code-coupling.

mohanchen replied:
If rdmft inherits esolver_ks_lcao, and then we just have a new esolver type?

maki49 replied:
Yes. After that, we can delete this new parameter.

mohanchen commented on source/module_rdmft/rdmft.cpp:
do not write a code > 500 lines

maki49 replied:
I think the length is related to his one-line-one-variable coding style -- that is what you want.

mohanchen replied:
Even if he writes the code in one-line-one-variable coding stype, it can still be controlled under 500 lines if you create more files.

JGHan7 replied:
OK, I will move the interface functions for getting each potential matrix to the newly created cal_V_rdmft.cpp,

move the functions for updating the system state to the newly created update_state_rdmft.cpp

mohanchen commented on tests/PP_ORB/Si_gga_7au_100Ry_2s2p1d.orb:
can we use available orbital files instead of adding a new one?

JGHan7 replied:
ok, I will use the old one

JGHan7 committed:
save

JGHan7 committed:
delete the new ord_file in tests/

JGHan7 committed:
last save befor move cal_V_ TV/hartree/XC from rdmft.cpp to cal_V_rdmft.cpp

JGHan7 committed:
last save before move the functions about update_state from rdmft.cpp to update_state_rdmft.cpp

JGHan7 committed:
save

JGHan7 committed:
save

JGHan7 committed:
Merge branch 'develop' of https://gitee.com/deepmodeling/abacus-develop into rdmft

JGHan7 committed:
prepare for PR

JGHan7 committed:
Merge branch 'rdmft_PR' of github.com:JGHan7/abacus-develop into rdmft_PR

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

maki49 commented:
@mohanchen Now the new .orb file is removed, and rdmft.cpp is separated into shorter files. 

Since this PR does not implement a complete ESolver with optimization (but only an initial guess generated by KSDFT), I think we can merge this PR at first without the new option of `esolver_type`. The new ESolver for RDMFT will be added in the future (maybe by Professor Kai Luo), then the newly added parameter can be deleted.

mohanchen commented:
> @mohanchen Now the new .orb file is removed, and rdmft.cpp is separated into shorter files.
> 
> Since this PR does not implement a complete ESolver with optimization (but only an initial guess generated by KSDFT), I think we can merge this PR at first without the new option of `esolver_type`. The new ESolver for RDMFT will be added in the future (maybe by Professor Kai Luo), then the newly added parameter can be deleted.

I see what you mean, but I need to review again the codes.

mohanchen committed:
Merge branch 'develop' into rdmft_PR

JGHan7 committed:
save

mohanchen commented on source/module_esolver/esolver_ks_lcao.cpp:
```diff
@@ -256,7 +262,15 @@ void ESolver_KS_LCAO<TK, TR>::before_all_runners(const Input_para& inp, UnitCell
-    // 15) if kpar is not divisible by nks, print a warning
+
+    // 15) initialize rdmft, added by jghan
+    if( PARAM.inp.ab_initio_type == "rdmft" )
```
if rdmft == true

JGHan7 replied:
Okay, Professor Chen

mohanchen commented on source/module_esolver/esolver_ks_lcao.cpp:
```diff
@@ -1068,6 +1100,37 @@ void ESolver_KS_LCAO<TK, TR>::iter_finish(int& iter)
     }
+
+    // 7) rdmft, added by jghan, 2024-10-25
+    if ( PARAM.inp.ab_initio_type == "rdmft" )
```
leaving approximately 5 lines here in esolver, I would suggest not adding any detailed operations in esolver_ks_lcao, the file is already very big.

JGHan7 replied:
esolver_ks_lcao is really too big... But I can't just change the first parameter of update_elec() from occNum to wg just to shorten the number of lines. This will be very indirect for future optimization of occupancy numbers. That means I have to multiply the occupancy number by wg every time I optimize it and then feed it to update_elec()

But the second half of the code about gradient and total energy will eventually be deleted. This is just for the verification calculation of obtaining the initial value smoothly. In the end, there can be only six or seven lines of code here

Is this solution acceptable?

JGHan7 replied:
correct： ”multiply the occupancy number by wk to get wg“, in the first paragraph

mohanchen commented on source/module_esolver/esolver_ks_lcao.cpp:
```diff
@@ -1162,6 +1225,29 @@ void ESolver_KS_LCAO<TK, TR>::after_scf(const int istep)
+    /******** test RDMFT *********/
+
+    // rdmft, added by jghan, 2024-10-17
+    if ( PARAM.inp.ab_initio_type == "rdmft" )
```
leaving less than 5 lines here

JGHan7 replied:
In next PR，I would delete this code

mohanchen commented on source/module_rdmft/rdmft.h:
```diff
@@ -0,0 +1,252 @@
+
+
+template <typename TK, typename TR>
+class RDMFT
```
It's not a good design for RDMFT because too many variables are defined here. I will not asking for modification this time since by now it will not affect other codes of ABACUS. However, it is possible that in future I need the developer to refactor these codes. Mark here.

JGHan7 replied:
RDMFT does have too many variables, I can accept it being refactored in the future

mohanchen commented on source/module_rdmft/rdmft_tools.cpp:
```diff
@@ -0,0 +1,433 @@
+
+
+// do wk*g(occNum)*wfcHwfc and add for TV, hartree, XC. This function just use once, so it can be replace and delete
+void add_wfcHwfc(const ModuleBase::matrix& wg, const ModuleBase::matrix& wk_fun_occNum, const ModuleBase::matrix& wfcHwfc_TV_in, const ModuleBase::matrix& wfcHwfc_hartree_in,
```
I would suggest one variable for a line

JGHan7 replied:
OK

mohanchen commented on source/module_rdmft/rdmft_tools.cpp:
```diff
@@ -0,0 +1,433 @@
+}
+
+
+
```
too many blank lines here.

JGHan7 replied:
I would delete it

mohanchen commented on source/module_rdmft/rdmft_tools.cpp:
```diff
@@ -0,0 +1,433 @@
+    for (int iat1 = 0; iat1 < ucell_in->nat; iat1++)
+    {
+        auto tau1 = ucell_in->get_tau(iat1);
+        int T1, I1;
```
initialize the variables.

JGHan7 replied:
OK

mohanchen commented on source/module_rdmft/update_state_rdmft.cpp:
```diff
@@ -0,0 +1,200 @@
+    }
+    /*********** what's this? When we use PBE-functional, this can't be deleted *************/
+
+    // // what this?
```
delete useless codes?

JGHan7 replied:
OK

JGHan7 committed:
modify ab_initio_type to bool rdmft

JGHan7 committed:
Merge branch 'develop' of https://gitee.com/deepmodeling/abacus-develop into rdmft

JGHan7 committed:
save

JGHan7 committed:
Fix: let nk_total *= nspin, in rdmft

JGHan7 committed:
modified according to Mr. Chen's suggestion

JGHan7 committed:
save

JGHan7 committed:
prepare for PR

JGHan7 committed:
Merge branch 'rdmft_PR' of github.com:JGHan7/abacus-develop into rdmft_PR

mohanchen commented on source/module_rdmft/cal_V_rdmft.cpp:
```diff
@@ -0,0 +1,339 @@
+            orb->cutoffs(),
+            &GlobalC::GridD,
+            nspin,
+
```
one extra line can be deleted

mohanchen commented on source/module_rdmft/cal_V_rdmft.cpp:
```diff
@@ -0,0 +1,339 @@
+            orb->cutoffs(),
+            &GlobalC::GridD,
+            nspin,
+
```
one extra line found

mohanchen commented on source/module_rdmft/cal_V_rdmft.cpp:
```diff
@@ -0,0 +1,339 @@
+                orb->cutoffs(),
+                &GlobalC::GridD,
+                nspin,
+
```
several similar places found (one blank line), please check

mohanchen commented on source/module_rdmft/rdmft.h:
```diff
@@ -0,0 +1,249 @@
+template <typename TK, typename TR>
+class RDMFT
+{
+  public:
```
do we need so many public variables and public functions, can we made some of  them as private?

JGHan7 replied:
Of course, in the future, more variables will be designed to be private based on actual conditions. But now our optimization plan is still being written, and more public variables are more convenient during debugging and testing.

mohanchen commented on source/module_rdmft/CMakeLists.txt:
I would suggest moving module_rdmft into module_hamilt_lcao， in future， maybe we will rename the module_hamilt_lcao to module_lcao_algorithms

JGHan7 replied:
I am currently writing programs to optimize the single-particle reduced density matrix，which will be in module_rdmft for now, and it is obviously not appropriate to put the optimized code in module_hamilt_lcao.

After all, this is just the production code for the rdmft method, and we still have a lot to add. Even if this PR fully meets the requirements, the new code will immediately destroy the current structure, making the movement of this detail useless.

mohanchen commented on source/module_rdmft/rdmft_tools.h:
```diff
@@ -0,0 +1,460 @@
+    #ifdef _OPENMP
+    #pragma omp parallel for schedule(static, 1024)
+    #endif
+    
```
one extra line found here

mohanchen commented on source/module_rdmft/rdmft_tools.h:
```diff
@@ -0,0 +1,460 @@
+
+
+template <>
+void conj_psi<double>(psi::Psi<double>& wfc);
```
It's not a good idea to 'invent' several new operations on Psi, these are common tools, not the tools by RDMFT itself.

JGHan7 replied:
I didn't find the corresponding operation for psi, so I wrote this function temporarily for the convenience of development. If psi can have this member function, it can be deleted immediately

kirk0830 replied:
I second this.

mohanchen commented on source/module_rdmft/rdmft_tools.h:
```diff
@@ -0,0 +1,460 @@
+          orb_cutoff_(orb_cutoff),
+          pot(pot_in),
+          hamilt::OperatorLCAO<TK, TR>(hsk_in, kvec_d_in, hR_in),
+
```
one extra space here

mohanchen committed:
Merge branch 'develop' into rdmft_PR

JGHan7 commented:
Most of the spaces here are because I modified the veff class according to my needs. I use a line of spaces to distinguish what I added myself, to remind me that a more suitable interface may be provided to me in the future, making the veff_rdmft class more concise or even deleting it.

JGHan7 committed:
save

JGHan7 committed:
merge the latest abacus

JGHan7 committed:
merge the latest abacus

JGHan7 committed:
merge the latest abacus

JGHan7 committed:
Merge branch 'develop' of https://gitee.com/deepmodeling/abacus-develop into rdmft

JGHan7 committed:
prepare for PR

JGHan7 committed:
Merge branch 'rdmft_PR' of github.com:JGHan7/abacus-develop into rdmft_PR

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

JGHan7 committed:
save

JGHan7 committed:
merge the latest abacus

JGHan7 committed:
add some tests for rdmft

JGHan7 committed:
Merge branch 'develop' of https://gitee.com/deepmodeling/abacus-develop into rdmft

JGHan7 committed:
save

JGHan7 committed:
prepare for PR

JGHan7 committed:
Merge branch 'rdmft_PR' of github.com:JGHan7/abacus-develop into rdmft_PR

JGHan7 committed:
save

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

JGHan7 committed:
reduce calculation time of tests: rdmft

JGHan7 committed:
modify some comments

JGHan7 committed:
Merge branch 'develop' of https://gitee.com/deepmodeling/abacus-develop into rdmft

JGHan7 committed:
prepare for PR

JGHan7 committed:
Merge branch 'rdmft_PR' of github.com:JGHan7/abacus-develop into rdmft_PR

JGHan7 committed:
modify tests

JGHan7 committed:
save

JGHan7 committed:
prepare for PR

JGHan7 commented:
This PR is ready to be merged, Professor Chen. 
Is there anything else that needs to be modified?@mohanchen

mohanchen commented:
> This PR is ready to be merged, Professor Chen. Is there anything else that needs to be modified?@mohanchen

I will review it

JGHan7 committed:
Merge branch 'develop' of https://gitee.com/deepmodeling/abacus-develop into rdmft

JGHan7 committed:
modify class RDMFT

JGHan7 committed:
save

JGHan7 committed:
modify Comments

JGHan7 committed:
Merge remote-tracking branch 'upstream/develop' into rdmft

JGHan7 committed:
save

JGHan7 committed:
prepare for PR

JGHan7 committed:
save

JGHan7 committed:
Merge branch 'develop' of https://gitee.com/deepmodeling/abacus-develop into rdmft

JGHan7 committed:
save

JGHan7 committed:
Merge branch 'rdmft' of https://gitlab.com/JGHan7/rdmft-abacus into rdmft

JGHan7 committed:
prepare for PR

JGHan7 committed:
save

JGHan7 committed:
Merge branch 'rdmft' into rdmft_PR

WHUweiqingzhou committed:
Merge branch 'develop' into rdmft_PR

mohanchen commented:
@kirk0830 do you have any more comments?

kirk0830 commented on docs/advanced/input_files/input-main.md:
```diff
@@ -4031,4 +4034,21 @@ The output files are `OUT.${suffix}/Excitation_Energy.dat` and `OUT.${suffix}/Ex
+ab-initio methods and the xc-functional parameters used in RDMFT.
+The physical quantities that RDMFT temporarily expects to output are the kinetic energy, total energy, and 1-RDM of the system in the ground state, etc.
+
+### rdmft
```
hear from @maki49 , present implementation of RDMFT still need something calculated with typical KS-DFT. Can you draw a schematic figure in the illustration of this PR, explaining the design of workflow? Say it might be like:

KS-DFT - (pass something) -> RDMFT.

This will be quite helpful for all reviewers understand your efforts with more details

JGHan7 replied:
I'm very sorry for not providing a clearer document.

Unlike DFT, RDMFT not only needs to optimize orbitals, but also needs to optimize occupation numbers. So far, in our program, the only relationship between RDMFT and KS-DFT is that wfc and occupation numbers wg/wk in atomic orbital basis after one iteration of KS-DFT are needed as initial values

(Of course, you can also use random numbers as initial values, so that RDMFT has nothing to do with KS-DFT. But we think the former may be a better solution. But at the request of Professor Chen, the code for getting the initial value in the iter_finish() function of ESolver_KS_LCAO has been deleted. The code in my new development branch has also inherited ESolver_KS_LCAO, and I think the next PR will have a clearer structure)

To be precise, all other variables passed in actually come from "ESolver_LCAO" instead of ESolver_KS_LCAO. I plan to draw a complete schematic diagram in the next PR. The current code is not complete enough and not easy to explain.

JGHan7 replied:
![image](https://github.com/user-attachments/assets/b4158a10-c55c-4c7a-ae08-7b0c3147f814)
This is a complete flow chart that can be used. The red is what I want from ABACUS, the green is what the current code gets, and the blue is what they are used for.

JGHan7 replied:
plus, n = eta = occupation numbers

kirk0830 commented on source/Makefile.Objects:
```diff
@@ -727,3 +729,8 @@ OBJS_TENSOR=tensor.o\
+
+OBJS_RDMFT=rdmft.o\
+        rdmft_tools.o\
+        cal_V_rdmft.o\
```
I hope your PR can be quickly understood by more developers. What is the "V" represents? There is a difference in the way of naming files, classes, functions and variables. Think this (here I use another example):
1. cal_H
2. cal_hamilt
3. cal_hmat

among them, actually the second is the most recommended, because:
1. there are other physical quantities represented by letter H, such as, enthalpy, the strength of magnetic field in matters.
2. "hmat" is both worse and better, what is worse is there is redundancy in name because the data structure will be indicated in the function's parameter list, but on the other hand what is better it provides the technical information (the data structure), but other developers should still find what is the meaning of "h".

Thanks.

JGHan7 replied:
I should thank you for your careful review and suggestions.

I mean, Hamilt_mat = V_T + V_nonlocal + V_local + V_hartree + V_xc_from_dft + V_special_exx.

In fact, from a physical point of view, I would prefer to change "V" to the name of "potential matrix (in NAOs)", but "potential" seems to have other meanings in ABACUS, so I just used "V".

I can change V to hamilt. But to be honest, I'm not sure if this is better.

JGHan7 replied:
The comment here was in response to the flowchart, but apparently I got it in the wrong place again...
sorry, sorry!

kirk0830 replied:
no, I mean, if you can only name the contribution term from dmft as "potential", then you can use term "pot_rdmft" as the file name. However, if the potential can be named more exactly, then please choose the best one.

JGHan7 replied:
Thank you, your suggestion and motivation are very clear. I have changed the file name to rdmft_cal_hamilt.cpp, but I suddenly realized that this may not be so appropriate.

Usually in KS-DFT, we add up the potential matrices to get the effective Hamiltonian H^KS of the KS system (that is, Hamilt_mat = V_T + V_local + V_hartree + ... as I wrote above), but in rdmft, there is no such universal and concise effective Hamiltonian H^RDMFT. For example, when using the CG method of Riemann space to optimize {φi}, we do not need to construct an effective Hamiltonian; when using an iterative diagonalization method, we need to artificially construct a Fock-like matrix (that is, the F matrix in the flowchart I posted), which is not a simple addition of V_T and V_hartree, and its eigenvalues ​​do not have the meaning of energy levels. So, from these perspectives, it is not appropriate to change the file name to hamilt.

I think the names of V_hartree and V_local are clear enough and consistent with the common symbols of physical formulas for each contribution term, and I want to continue using them. So should I continue to use the name rdmft_cal_V.cpp and add more detailed comments to the function declaration, or use the name rdmft_cal_pot.cpp?

Or do you have any other suggestions?

JGHan7 replied:
Thank you, your suggestion and motivation are very clear. I have changed the file name to rdmft_cal_hamilt.cpp, but I suddenly realized that this may not be so appropriate.

Usually in KS-DFT, we add up the potential matrices to get the effective Hamiltonian H^KS of the KS system (that is, Hamilt_mat = V_T + V_local + V_hartree + ... as I wrote above), but in rdmft, there is no such universal and concise effective Hamiltonian H^RDMFT. For example, when using the CG method of Riemann space to optimize {φi}, we do not need to construct an effective Hamiltonian; when using an iterative diagonalization method, we need to artificially construct a Fock-like matrix (that is, the F matrix in the flowchart I posted), which is not a simple addition of V_T and V_hartree, and its eigenvalues ​​do not have the meaning of energy levels. So, from these perspectives, it is not appropriate to change the file name to hamilt.

I think the names of V_hartree and V_local are clear enough and consistent with the common symbols of physical formulas for each contribution term, and I want to continue using them. So should I continue to use the name rdmft_cal_V.cpp and add more detailed comments at the beginning of the file, or use the name rdmft_cal_pot.cpp?

Or do you have any other suggestions?

kirk0830 replied:
@JGHan7 
hmmm...I am not sure. 
For the file name, I think "cal_" always appears in the function name, which indicates "this is the function **calculates** something". This is the reason why I suggest the term "pot_rdmft", you can place several related implementation in this file, and there will be not any misleading words.

For the concept of "potential", there is a folder named source/module_elecstate/potentials, several potential implementations are placed there. Could you check if that fits your case? In that folder, you will find there are files named starting with "pot_".

But you are the one knows the most about your codes, so you can make your decision.

JGHan7 replied:
What you said makes sense, I will delete "cal_".

The potential under source/module_elecstate/potentials seems to be the potential obtained by grid integration, which is consistent with what I have here in the physical sense, but I want to emphasize that under the NAOs representation, perhaps the name rdmft_pot_mat can be used

kirk0830 replied:
wait, mat is the data structure, it would be better not include it in file name

JGHan7 replied:
OK, I use rdmft_pot.cpp to instead

kirk0830 commented on source/module_esolver/esolver_ks_lcao.cpp:
```diff
@@ -62,6 +62,11 @@
 
+// test RDMFT
+#include "module_rdmft/rdmft.h"
+#include "module_elecstate/module_dm/cal_dm_psi.h"  //temp ,delete
```
I notice you annotated "temp, delete". Is it the time that you can delete it?

JGHan7 replied:
Thanks for reminding me, I'll delete it now.

kirk0830 commented on source/module_esolver/esolver_ks_lcao.cpp:
```diff
@@ -1081,6 +1094,23 @@ void ESolver_KS_LCAO<TK, TR>::after_scf(const int istep)
     ModuleBase::timer::tick("ESolver_KS_LCAO", "out_deepks_labels");
 #endif
 
+    /******** test RDMFT *********/
```
is the following code piece planned to be deleted before merging this PR?

JGHan7 replied:
Currently this is the only code that can produce results and complete integration tests, so it cannot be deleted yet.

kirk0830 commented on source/module_hamilt_general/module_xc/xc_functional.cpp:
```diff
@@ -18,6 +18,7 @@ std::vector<int> XC_Functional::func_id(1);
 int XC_Functional::func_type = 0;
 bool XC_Functional::use_libxc = true;
 double XC_Functional::hybrid_alpha = 0.25;
+std::map<int, double> XC_Functional::scaling_factor_xc = { {1, 1.0} }; // added by jghan, 2024-10-10
```
this is a very interesting extension of XC. Actually, this factor is of need more than RDMFT, but due to presently there is no user claims he/she needs this so there is nobody implements.

Thanks for your contribution.

kirk0830 commented on source/module_hamilt_general/module_xc/xc_functional.cpp:
```diff
@@ -196,13 +198,68 @@ void XC_Functional::set_xc_type(const std::string xc_func_in)
+        func_type = 4;
+        use_libxc = true;
+    }
+    // just for test BLYP in libxc, added by jghan, 2024-10-22
```
BLYP is a common XC, it is a little bit pity that ABACUS did not implement it as a built-in XC. But I am curious why you add the suffix "_LIBXC"?

kirk0830 replied:
plus, you annotated "just for test", does it mean you plan to delete this option before merging this PR?

JGHan7 replied:
Sorry, I added "_LIBXC" because I didn't check carefully whether ABACUS implements built-in BLYP, and I'm not sure whether BLYP (from libxc) will be approved.

In fact, I implemented BLYP here for detailed debugging, but since ABACUS's long-range exx code still has convergence problems (I know someone in the ABACUS team is solving this problem, but unfortunately the code has not been submitted yet), debugging is not possible yet.

I can remove the suffix "_LIBXC" or delete this code. What do you prefer?

JGHan7 replied:
I checked and found that ABACUS has a built-in BLYP. 
When I implemented it, I was worried that the results of the built-in BLYP in ABACUS and the BLYP in libxc would be different, so I used this when debugging. I have now deleted it.

kirk0830 commented on source/module_hamilt_general/module_xc/xc_functional_libxc_vxc.cpp:
```diff
@@ -240,10 +252,12 @@ std::tuple<double,double,ModuleBase::matrix,ModuleBase::matrix> XC_Functional_Li
         for( int ir=0; ir<nrxx; ++ir )
         {
-            if ( rho[ir*2]<rho_th || sqrt(std::abs(sigma[ir*3]))<grho_th || std::abs(kin_r[ir*2])<tau_th)
+            if ( rho[ir*2]<rho_th || sqrt(std::abs(sigma[ir*3]))<grho_th || std::abs(kin_r[ir*2])<tau_th) {
```
if I understand correctly, the if conditional statement actually handles three different conditions. For readability, my suggestion would be:

const bool "the first condition" = rho[ir*2] < rho_th;
const bool "the second condition" = sqrt(std::abs(sigma[ir*3])) < grho_th;
...

then the if would be:
if ("the first condition"||"the second..."||...)

how do you think?

JGHan7 replied:
I didn't write this code, and I don't know what it does. I guess the reason it's edited is because the GitHub bot added "{}" to it.

kirk0830 commented on source/module_hamilt_general/module_xc/xc_functional_libxc_vxc.cpp:
```diff
@@ -240,10 +252,12 @@ std::tuple<double,double,ModuleBase::matrix,ModuleBase::matrix> XC_Functional_Li
                 sgn[ir*2] = 0.0;
-            if ( rho[ir*2+1]<rho_th || sqrt(std::abs(sigma[ir*3+2]))<grho_th || std::abs(kin_r[ir*2+1])<tau_th)
+}
+            if ( rho[ir*2+1]<rho_th || sqrt(std::abs(sigma[ir*3+2]))<grho_th || std::abs(kin_r[ir*2+1])<tau_th) {
```
above comment also applies to this

kirk0830 commented on source/module_rdmft/cal_V_rdmft.cpp:
```diff
@@ -0,0 +1,340 @@
+        elecstate::psiMulPsiMpi(wk_funEta_wfc, wfc, DM_Kpointer, ParaV->desc_wfc, ParaV->desc);
+#else
+        elecstate::psiMulPsi(wk_funEta_wfc, wfc, DM_Kpointer);
+#endif            
```
I think it may be not directly feasible but I would like to share a more skilled idea of writhing this piece of code.

Ideally, you should have only one function named "a", instead of two functions "a" and "a_mpi". In the function "a", you implement those conditional codes.

But it might be okay now, because I am not sure if this is really possible for present code structure.

JGHan7 replied:
I understand what you mean, I will try to use a function name in my code. But the function here is not written by me, I just use it

kirk0830 commented on source/module_rdmft/cal_V_rdmft.cpp:
```diff
@@ -0,0 +1,340 @@
+void RDMFT<TK, TR>::get_DM_XC(std::vector< std::vector<TK> >& DM_XC)
+{
+    // get wk_funEta_wfc = wk*g(eta)*conj(wfc)
+    psi::Psi<TK> wk_funEta_wfc(wfc);
```
does the `wk_funEta_wfc` operation have a more physical/realistic name?

JGHan7 replied:
I can't think of a better name, this variable just makes the calculation process more convenient. The name of the variable is simply obtained through the calculation process

kirk0830 commented on source/module_rdmft/cal_V_rdmft.cpp:
```diff
@@ -0,0 +1,340 @@
+    }
+
+    // update HR_TV in ion-step, now HR_TV has the HR of V_ekinetic + V_nonlcao + V_local
+    V_ekinetic_potential->contributeHR();
```
is the function `contributeHR` called here related with the one which looks quite similar in ABACUS LCAO implementation?

JGHan7 replied:
Sorry, I don't quite understand what you mean. 
This function is indeed related to ABACUS LCAO, because it is a class member function provided by the ABACUS interface.
![1732371009709](https://github.com/user-attachments/assets/1162190a-16ae-49f8-bdeb-c788809def9b)

kirk0830 commented on source/module_rdmft/rdmft.h:
```diff
@@ -0,0 +1,200 @@
+
+    // natrual occupation numbers and wavefunction
+    ModuleBase::matrix occ_number;
+    psi::Psi<TK> wfc;
```
this is a good practice: once possible, name variable with physical name instead of some symbol.

JGHan7 replied:
Thanks, I will try to name it according to the physical meaning

kirk0830 commented on source/module_rdmft/rdmft.h:
```diff
@@ -0,0 +1,200 @@
+    //! construct V_XC based on different XC_functional( i.e. RDMFT class member XC_func_rdmft)
+    void cal_V_XC();
+
+    double cal_E_gradient();
```
what is the meaning of gradient here? it is gradient of E but respect to what?

JGHan7 replied:
get the gradient of E with respect to wave function and occupation number respectively. I think this name is too long, so I just specify the gradient.

kirk0830 commented on source/module_rdmft/rdmft_tools.cpp:
```diff
@@ -0,0 +1,414 @@
+
+
+template <>
+void psiDotPsi<double>(const Parallel_Orbitals* ParaV, const Parallel_2D& para_Eij_in,
```
I will not comment on your code structure, but care much about readability. As far as I understood, this function actually does the inner product of two Hilbert vector in representation of some basis, the one is psi and the other might be the psi after act by some operator? So it is possible to pass H|psi> to wfc, and |psi> to H_wfc? also the "H" is not directly/specifically the Hamiltonian operator, but actually any operator?

JGHan7 replied:
Your understanding of this function is correct, but I don't quite understand what you meant by the second half of your sentence.

Here H can be any operator, and I originally thought that the parameter Dmn here was useless, and I was considering how to delete it in the future. But now I find that the optimization plan I am currently working on requires this variable, so I didn't reconstruct this function

So the inner product here is a bit special. For example, for wfc with the same k points, we have
sum_mu conj( wfc(k, m, mu) )* H_wfc(k, n, mu) = Dmn(k)
And the current code only shows that wfcHwfc = Dnn(k) is useful

This function name is indeed not clear enough. Do you have a better naming scheme?

kirk0830 replied:
seems you are doing two things in one function, the first is to calculate the inner product of bra and op_ket, the other is to extract diagonal elements in bra_op_ket. So the function you named PsiDotPsi is a quite special function that should be properly treated. My suggestion is, you can separate your function into two smaller functions, the first named cal_bra_op_ket, and the second named _diagonal_in_serial. Then you get the wfcHwfc by _diagonal_in_serial(cal_bra_op_ket(..., bra, ..., op_ket, ...)).
Do you think this is better?
You can then add unittests on them to ensure the correctness and stability.

JGHan7 replied:
Yes, your solution is better, I will modify it.
But considering that nbands will be relatively large in the RDMFT method, I still plan to use the formal parameter Dmn(k) of the reference variable instead of the return value in the function cal_bra_op_ket(…), is this okay?
Also, I am a little curious about why you used "_" as the first letter when naming _diagonal_in_serial()

kirk0830 replied:
@JGHan7 I think it is acceptable.
the reason why I add one "_" ahead is to indicate this is a private or context-specific implementation (not designed to be generally used elsewhere). Now there are new implementations combining this nomenclature with anonymous namespace (a relatively advanced C++ feature), to make the interface of code very clean. For an instance, please see: 
https://github.com/deepmodeling/abacus-develop/blob/6df5c6dc254c204f8d6acfa7f01cf5f4de2ca1a0/source/module_base/grid/batch.cpp#L41

JGHan7 replied:
OK, thank you for your detailed explanation.

kirk0830 commented on source/module_rdmft/rdmft_tools.cpp:
```diff
@@ -0,0 +1,414 @@
+    } else if ( symbol==5 ) { return 1.0;
+    } else 
+    {
+        std::cout << "\n!!!!!!\nThere may be some errors when calling wg_fun()\n!!!!!!\n";
```
what is the behavior you really expect here? Do you want to trigger an assertion failure here? if so, you use assert(symbol <= 5) instead.

JGHan7 replied:
OK, I'll modify it

kirk0830 commented on source/module_rdmft/rdmft_tools.h:
```diff
@@ -0,0 +1,429 @@
+
+
+
+namespace rdmft
```
this is quite a good practice compared with other codes because you use namespace instead of any absurd class to place tool functions.

kirk0830 reviewed (commented):
I have reviewed this PR. Thanks for your contribution, this PR brings new functionality to ABACUS. 
However, due to I am not expert in RDMFT, EXX, so the one I care most in this PR is the code readability. I hope from my comments you will get more programming experience, also I hope your code can be understood by more developers more deeply and quickly.

Thanks.

JGHan7 commented:
@kirk0830 Thank you for your professional comments. I sincerely look forward to learning more about code from you.

JGHan7 committed:
save

JGHan7 committed:
Merge branch 'develop' of https://gitee.com/deepmodeling/abacus-develop into rdmft

JGHan7 committed:
Merge branch 'rdmft' into rdmft_PR

JGHan7 committed:
Merge branch 'rdmft_PR' of github.com:JGHan7/abacus-develop into rdmft_PR

JGHan7 committed:
modify

JGHan7 committed:
Merge branch 'develop' of https://gitee.com/deepmodeling/abacus-develop into rdmft

JGHan7 committed:
Merge branch 'rdmft' into rdmft_PR

JGHan7 committed:
save

JGHan7 commented:
I'm so sorry. Since I don't know how to use GitHub, all my replies in the past month are in "pending" status, and only I can see them (I wonder why no one has ever replied to me...). 
I'm really sorry, everyone. I will learn quickly and at least successfully push the replies from the past few days.

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

kirk0830 reviewed (commented):
Hi, a small change request is made, please check if my suggestion is correct, Thanks

JGHan7 committed:
add a comment in rdmft_cal_hamilt/V.cpp

JGHan7 committed:
rename rdmft_cal_hamilt.cpp to rdmft_pot_mat.cpp

JGHan7 committed:
Refactore psiDotPsi()

JGHan7 committed:
save

JGHan7 committed:
Merge branch 'develop' of https://gitee.com/deepmodeling/abacus-develop into rdmft

JGHan7 committed:
save

JGHan7 committed:
save

JGHan7 committed:
Merge remote-tracking branch 'upstream/develop' into rdmft

JGHan7 committed:
Merge branch 'rdmft' into rdmft_PR

JGHan7 committed:
Merge branch 'rdmft_PR' of github.com:JGHan7/abacus-develop into rdmft_PR

JGHan7 committed:
save

JGHan7 committed:
Merge branch 'develop' of https://gitee.com/deepmodeling/abacus-develop into rdmft

JGHan7 committed:
Merge branch 'rdmft' into rdmft_PR

mohanchen committed:
Merge branch 'develop' into rdmft_PR
